### PR TITLE
[BD-21] v1.2.0 Deprecate waffle namespaces (in a backward-compatible manner)

### PR DESCRIPTION
**Pull request by regisb**
**Description:** Start the deprecation process of WaffleFlagNamespace and WaffleSwitchNamespace. All related features and methods are displaced to the WaffleFlag/WaffleSwitch classes. We still expose the legacy classes to verify that the IDAs are compatible with them.

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation

**Reviewers:**
- [x] @robrap 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-656](https://openedx.atlassian.net/browse/BLENDED-656) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


When this pull request is ready, tag your edX technical lead.

**Comment by regisb** [comment]
This is ready for review @robrap.

File: edx_toggles/toggles/__init__.py
Comment (robrap) [inline edx_toggles/toggles/__init__.py:5]: To enable expand/contract, I was thinking this version would contain both the current names and the legacy names, pointing to the same classes. This would enable the introduction of the legacy names in advance of breaking compatibility for the existing names, to remove the need for lock-step merges in the future.

I don't know if this code works, or if there is another way you'll need to import these, but something like:
```
LegacyWaffleFlag = WaffleFlag
LegacyWaffleFlagNamespace = LegacyWaffleFlagNamespace
LegacyWaffleSwitch = WaffleSwitch
LegacyWaffleSwitchNamespace = WaffleSwitchNamespace
```

Comment (regisb) [reply edx_toggles/toggles/__init__.py:5]: Right. I understand.

Comment (robrap) [reply edx_toggles/toggles/__init__.py:5]: May be obvious, but please add a comment as well.  Thanks.

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [inline edx_toggles/tests/test_waffle.py:21]: 1. I thought you needed to use as context?  For example:
```
        with self.assertRaises(
            ValueError, WaffleFlag, "non_namespaced", module_name="module1"
        ):
            WaffleFlag("namespaced.name", module_name="module1")
```
2. This might be passing as-is (if it is) because you are in fact sending it a namespaced name.

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [inline edx_toggles/tests/test_waffle.py:12]: Ultimately, is this the only planned coverage? Do we need a TODO to create or move tests?

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [inline edx_toggles/tests/test_waffle.py:20]: Should we include the flag name, or do you think the stacktrace will be enough?

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [inline edx_toggles/tests/test_waffle.py:30]: Same questions as with WaffleFlagTest.

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [inline edx_toggles/tests/test_waffle.py:35]: Should this be on WaffleFlag as well?

File: edx_toggles/toggles/testutils.py
Comment (robrap) [inline edx_toggles/toggles/testutils.py:71]: I didn't noticed this earlier because I wasn't looking at `override_waffle_flag` and `override_waffle_switch` code side-by-side, but why are these implemented differently?
* Is there an actual difference?
* If not, could we choose your favorite implementation and duplicate it?
* Can we get the docs more consistent and complete here?
This could be a follow-up if it isn't quick.

File: edx_toggles/toggles/testutils.py
Comment (robrap) [inline edx_toggles/toggles/testutils.py:77]: Reminder: Can we use `_cached_switches` directly so we can deprecate/remove `set_request_cache`?

**Review by robrap — COMMENTED** [review]
Thanks.  If this is going to take some time to land, you may want to consider a PR that simply adds the following with no other changes in a separate PR:
```
LegacyWaffleFlag = WaffleFlag
LegacyWaffleSwitch = WaffleSwitch
```
This would unblock PRs moving on to the newly named classes, and to allow the refactoring (which should be backward compatible) to come in this PR afterwards.

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/flag.py:95]: > Can we rename _monitor_value to set_monitor_value?

Yes, but then we need to keep the `_monitor_value` name in the legacy class, as it is used in edx-platform's CourseWaffleFlag.

I'll update the docstring and rename the method.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/base.py:16]: Well, it's possible, but I'm not a big fan. As I mentioned elsewhere, the `module_name` is only used to track code owners in the toggle state endpoint. The `module_name` is useless in the primary use case for feature toggles -- which is to actually enable/disable features. So I would be reluctant to make that argument mandatory.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/base.py:25]: Yes, I purposefully ignored that :innocent: I usually have a distaste towards boolean arguments that enable/disable a certain behaviour. This makes the code more intricate and usually implies that more edge cases need to be taken into account. (For instance, ExperimentWaffleFlag instances need the namespace in order to work properly.)

Boolean arguments can most often be replaced by good composition. For instance, in this particular case it's easy and more explicit to bypass the name check by creating a class that calls the BaseToggle constructor, but not the BaseWaffle constructor. 

Furthermore, all current waffles make use of namespaces, so we don't really need to bypass the name check, yet.

That being said, if you really want to stick to the ADR then I can implement the argument.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:59]: Yes, I think you're right; this might be too verbose. Also, it's made redundant by the fact that we will explicitly be relying on legacy classes.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:119]: No, not really. I'll remove them.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:60]: Yes -- I'll add the same argument here as in BaseToggle.

File: edx_toggles/toggles/internal/waffle/switch.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/switch.py:39]: We can simply get rid of those methods and preserve `set_request_cache_with_short_name` in the legacy class.

File: edx_toggles/toggles/__init__.py
Comment (regisb) [reply edx_toggles/toggles/__init__.py:11]: I think we should keep them; it makes it quite explicit that we are relying on legacy classes and it's much easier to grep.

File: edx_toggles/toggles/internal/waffle.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle.py:39]: Hmmm your remark made me realize that we would be breaking more than one thing. The LegacyWaffleFlag class no longer has a NAMESPACE_CLASS or a waffle_namespace attribute. I'll reintroduce these legacy attributes and opened an edx-platform PR to remove the calls to the internal API: https://github.com/edx/edx-platform/pull/25538

File: edx_toggles/toggles/internal/waffle/__init__.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/__init__.py:2]: Added as a follow-up task.

File: edx_toggles/tests/test_waffle.py
Comment (regisb) [reply edx_toggles/tests/test_waffle.py:21]: Actually, I never used `assertRaises` as a context. It is possible to pass a callable and its arguments to assertRaises: https://docs.python.org/3.7/library/unittest.html#unittest.TestCase.assertRaises
I like the context usage better, so I'll use it here.

File: edx_toggles/tests/test_waffle.py
Comment (regisb) [reply edx_toggles/tests/test_waffle.py:12]: I didn't see fit to introduce more tests, as most of them are already present in `test_legacy_waffle.py`. I guess the tests from test_legacy_waffle should be moved here once the legacy classes are removed.

File: edx_toggles/tests/test_waffle.py
Comment (regisb) [reply edx_toggles/tests/test_waffle.py:20]: The ValueError will include the non-namespaced flag name.

File: edx_toggles/tests/test_waffle.py
Comment (regisb) [reply edx_toggles/tests/test_waffle.py:35]: No, because they use the same constructor and we want to avoid too many tests crashing at the same time for the same reason.

File: edx_toggles/toggles/testutils.py
Comment (regisb) [reply edx_toggles/toggles/testutils.py:71]: > Is there an actual difference?

Yes. The `override_waffle_switch` makes an explicit call to `WaffleSwitch.is_enabled`, which fills the cache, while `override_waffle_flag` only looks at the cached values. At this point I'd rather avoid changing the behaviour of these methods, which are quite sensitive.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:104]: Related to an earlier discussion about protecting against backward-incompatible changes, can we add a property for this, and then add a custom attribute for this and for all the deprecated methods that will be going away.  

We can either go with something like:
```
deprecated_waffle_legacy_method=WaffleSwitch[some_name].switch_name
```
Or
```
deprecated_waffle_legacy_switch_name=some_name
```
It would be easier to query for the former, static key name, but it means we just get the last value, so it could be more like an onion, where we solve one issue and then find the next.

Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:104]: Sure, I can do either. Do you prefer we go with the former?

Comment (robrap) [reply edx_toggles/toggles/internal/waffle/legacy.py:104]: @regisb: We can absolutely follow-up with this if you want to land this first. ^  Just need it before 2.0.0.  I still need to double-check this PR once I am done with meetings.

Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py]: Please let me know if you agree with the changes I pushed.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/base.py:25]: Feel free to propose an alternate solution. The requirement is clear.  There is waffle usage in other IDAs, like https://github.com/edx/ecommerce/search?q=USER_LIST_VIEW_SWITCH, and I want them to be able to have a simple path to using the new classes without having to adjust the flag and switch names.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/base.py:16]: I feel reasonably confident that I want this for the following reasons:
* We need it to maintain our system, and it is the least expensive way to ensure it is there. Linting is a possibility, but it is a lot more complicated to make sure linting is happening everywhere you want it to.
* The toggle state endpoint reports the `module_name`, whether or not it is mapped to a code owner.  Could still be useful to anyone for tracking down toggles that come in from plugins, etc.
* Since `__name__` is always available, it basically puts no extra burden on developers.

If you still disagree, I'm happy to have us solicit additional opinions.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/base.py:25]: I propose to expose NonNamespacedWaffleFlag and NonNamespacedWaffleSwitch in the `__future__` API. I'll push my change to this PR.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:115]: I wouldn't change behavior and chance bugs.  Are there names with multiple `.`s?

Can we just adjust above to use:
```
        self._switch_name = switch_name
```
And the return `self._switch_name`?

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:114]: Great.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/legacy.py:115]: Also, please add for `flag_name` as well.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:200]: Typo in `wafle_namespace`

**Review by robrap — COMMENTED** [review]
These are just commit comments.  I still need to re-review the latest changes.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:115]: :ok_hand:

File: edx_toggles/toggles/testutils.py
Comment (robrap) [reply edx_toggles/toggles/testutils.py:70]: Makes sense not to touch right now.  Maybe a (potentially duplicated) docstring?
```
Note: The implementation for overriding switches and flags differ. The ``override_waffle_switch``
  makes an explicit call to ``WaffleSwitch.is_enabled``, which fills the cache, while
  ``override_waffle_flag`` only looks at the cached values. It is unclear whether this difference
  is important, or just due to being developed at different times by different people.
```
Maybe I over explain, but when I have a question on a PR, I often find myself re-asking  the same question in the future and not remembering the answer. :)

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [reply edx_toggles/tests/test_waffle.py:39]: Maybe a name like `test_base_constructor` or a comment would make that more clear.

Comment (regisb) [reply edx_toggles/tests/test_waffle.py:39]: I'll move this to test the base waffle class.

File: edx_toggles/tests/test_waffle.py
Comment (robrap) [reply edx_toggles/tests/test_waffle.py:12]: Can you add this as a task, or subtask, so that we don't drop coverage assuming the tests already exist?  It could also be a TODO  comment in the docstring of this file.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/base.py:38]: Typo: `namespaved`

File: edx_toggles/toggles/internal/waffle/base.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/base.py]: It's already pushed to this PR. Should I remove it?

File: edx_toggles/tests/test_waffle.py
Comment (regisb) [reply edx_toggles/tests/test_waffle.py:30]: I added a comment in the docstring of test_legacy_waffle.py.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/base.py:38]: Could you add the comment here that this is only intended to be used by child classes with different name validation requirements?

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/flag.py:99]: Second sentence is a mix of singular/plural that makes it a little hard to read.  I think switching `Child classes are...` to `A child class is..`, and fixing the rest to be singular, might read the best.

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/flag.py:104]: Great.  Can we do a quick edit to the ADR now or later to avoid confusion?

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/flag.py:157]: I was originally confused by this.  Maybe the description needs an example of setting the setting?  If it were a dict, the values would just be ignored, right?

File: edx_toggles/toggles/internal/waffle/switch.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/switch.py:14]: Reminder, unless I am re-reviewing too early. :)

File: edx_toggles/toggles/internal/waffle/switch.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/switch.py:21]: Please try to mark resolved after it is fixed on the PR.  Thanks.

**Comment by robrap** [comment]
1. Let me know when this is ready for review again.
2. Can you open and link to an edx-platform PR with this upgrade to ensure tests all pass?
Thanks.

File: edx_toggles/toggles/internal/waffle/base.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/base.py:38]: I already added it to `BaseToggle.validate_name` but I can add it here as well.

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/flag.py:157]: > Maybe the description needs an example of setting the setting?

I'll make the warning more precise.

> If it were a dict, the values would just be ignored, right?

Yes, absolutely.

**Comment by regisb** [comment]
Sorry about the incorrect resolved comments @robrap, somehow I made the changes in the `flag.py` file but not in `switch.py`.
I'm now rebasing on top of master to resolve conflicts, but you should only have to review the latest commit.

**Comment by regisb** [comment]
Here is the edx-platform PR that tests compatibility: https://github.com/edx/edx-platform/pull/25550
@robrap this is ready to review again.

**Review by robrap — APPROVED** [review]
Minor comments.  You can fix, squash, and merge, once the edx-platform PR is green: https://github.com/edx/edx-platform/pull/25550

**Comment by robrap** [comment]
@regisb: To make sure this is clear, I left comments that you can decide when and how to best handle them. This is approved and I’ll leave it to you to merge/release when you are ready. Thanks.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py:21]: I added the imports to the docstring. I think it makes more sense to keep this here in the legacy module, because this is where people will look first after they upgrade to 1.2.0.

**Comment by regisb** [comment]
@robrap I released [1.2.0](https://github.com/edx/edx-toggles/releases/tag/1.2.0) this morning. Any idea why it was not pushed to pypi yet? https://pypi.org/project/edx-toggles/

**Comment by robrap** [comment]
@jmbowman: Is it possible that Travis changes have stopped builds on tags?  I'm not clear on why there isn't a 1.2.0 build in Travis.

**Comment by jmbowman** [comment]
Travis changed our billing plan with very short notice to one that we exhausted less than 24 hours after it was changed, so there have been a couple of periods since then when Travis has refused to run any of our jobs as we try to navigate the new billing options.  It should be working again at the moment, if you want to try with a new tag.

**Pull request merged**
